### ISSUE TITLE
Configuring Kernel Websocket Ping timeout value during kernel launch.

### ISSUE THREAD

**Issue by IMAM9AIS** [jupyter/enterprise_gateway#696]
Hello,

Here is a brief background of the problem :- 

Jupyter Notebook uses https://github.com/jupyter/notebook/blob/master/notebook/base/zmqhandlers.py 
in-order to create and handle websocket connection with the client. These ZMQ Handlers have a ping pong mechanism to check the health of websocket connection between the ports of client and kernel machine. When a kernel is spawned remotely using JEG, any disconnect of network between client and remote kernel ( be it laptop going into sleep mode or internet disconnectivity) results in the zmqhandlers disconnecting the websocket connection between these ports(after 90s of timeout hardcoded in the code ). The users have no idea of this, since the state of the kernel does not reflect this. Although a user can refresh the jupyter lab page or reconnect to kernel (Command), this does not seem to good option, given a kernel can have processes running for a long time and can leave users clueless about it's results. 

Jupyter Notebook provides us with an option to toggle the ping pong mechanism or increase it's timeout value by providing parameter value as --NotebookApp.tornado_settings={'ws_ping_timeout'=somevalue} during notebook startup. 

However given that spawning of kernels is something that is happening remotely with JEG using notebook to create ZMQ ports, what should be the right way to set the value of this variable during a kernel launch ? 

PS:- We have tried modifying the timeout value that is hard coded in the zmqhandlers.py file and things reflect fine. But passing this value during the launch rather than having a hardcoded one seems a better approach .

**Comment by kevin-bates** [comment]
Not sure if this is applicable (not much of a web server dev), but @esevan contributed this https://github.com/jupyter/nb2kg/pull/29 a couple months ago.  Are you using an updated NB2KG?

**Comment by IMAM9AIS** [comment]
Hi Kevin,

We are using the latest version.I am not sure if this should fix our problem. The above commit is mainly to set timeouts for nginx proxy connections, while the websocket i am interested in is setting timeouts for ports associated with kernel launch for JEG's side(which seems to use notebook server zmqhandlers) and jupyter lab client. 

I believe there should be a way to set this param.

I tried doing this, 
jupyter enterprisegateway --ip=* --NotebookApp.tornado_settings="{'ws_ping_interval':36500}"

but it does not seem to work.

**Comment by kevin-bates** [comment]
Yeah, any `NotebookApp` traits won't be recognized since EG derives from Kernel Gateway's `KernelGatewayApp` which then derives from `JupyterApp`.

As a proof of concept, are you in a position to try adding something like the following after [this line](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/enterprisegatewayapp.py#L282)?

```python
self.web_app.settings['ws_ping_interval'] = 36500
```
If that works, or some version thereof, we can formalize it as a parameter.  You might find this change is required in Kernel Gateway instead.  So if this doesn't work, it might be worth altering Kernel Gateway's [`init_webapp()`](https://github.com/jupyter/kernel_gateway/blob/master/kernel_gateway/gatewayapp.py#L451).

**Comment by IMAM9AIS** [comment]
Oh Yes.
Setting the value of ping_interval from JEG's init function does the job for us.
We were actually thinking of going through this route. It seems this the route that was taken to set the value of 'allow_remote_access' variable too that occurs in settings.

Since the value of this variable is something that will vary on use-case basis (Some people might want to disable this feature while some might increase/decrease it), i think it would be best if we set this using the environment variables itself. 

Something like:-
self.web_app.settings['ws_ping_interval'] = os.getenv("WS_PING_INTERVAL",3000)

**Comment by kevin-bates** [comment]
Awesome - that's great news!  Let's go ahead and expose this as a command-line option that defaults itself from an env that then defaults _itself_ from `3000` (or whatever the proper default should be - 90 seconds is the implied default from above).

[This code](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/enterprisegatewayapp.py#L161-L171) is a reasonable example of what would be necessary, with `max_kernels_per_user` replaced with `ws_ping_interval`.

**Comment by esevan** [comment]
@IMAM9AIS Hi, IMAM. Could you please explain what's your scenario to use that variable?
As I understand, there're two websocket connections in EG design as below.
```
User -----ws1----- NB2KG -----ws2----- EG
```
Where I fixed was ws1 and where you're pointing out is ws2, isn't it?
I'm wondering what happens if ws2 timeout is 90s, because I've not seen any disconnection in my notebook after 90s since I applied https://github.com/jupyter/nb2kg/pull/29 patch.

It would be very appreciated to prevent unknown issues, thanks!

**Comment by IMAM9AIS** [comment]
@esevan Hi, 

I think you have rightly pointed out the web socket design in EG. 

I think my understanding was not correct as i didn't take into consideration NB2KG relay in between the websockets. 
The websocket connection that i am referring to(taking into consideration the above design), is the connection between ZMQ ports spawned during kernel launch(remote kernels) and the  NB2KG ports.

You can replicate this issue by launching a kernel remotely with a remote instance of EG, disconnecting the internet for approx 2 minutes and then reconnecting back. 

The EG will show logs as :- "WebSocket ping timeout after (some variable above 90s value)  ms". 
This is inherent to checks that are being performed in this file 
https://github.com/jupyter/notebook/blob/master/notebook/base/zmqhandlers.py
following which there is a websocket close event that is being performed by these handlers.

Any execution of code from the users end will print this error:- 
" Exception writing message to websocket:"

I hope this can help you understand and replicate this issue, following which the solution discussed above makes sense to proceed with .

**Comment by esevan** [comment]
@IMAM9AIS Thanks for your kind detail. Nice catch :+1:

**Comment by IMAM9AIS** [comment]
@kevin-bates can you please give me the relevant permission. I am not able to raise a PR .

**Comment by kevin-bates** [comment]
Anyone should be able to create a PR.  I'm not familiar with what else you'd need.  Copying @lresende who has rights to elevate privileges anyway, but I don't think there's anything we need to do.  

Ah, are you trying to open a PR directly to the repo?  If so, you need to create a fork and push your changes to a specific branch of the fork, then open the PR from that.  Not sure I have the all the terminology but I suspect you're attempting to open a PR on the repo directly.

**Comment by IMAM9AIS** [comment]
So sorry for the trouble. I had to create a fork first !
Here is the PR- https://github.com/jupyter/enterprise_gateway/pull/698

**Comment by esevan** [comment]
This seems to be resolved by #698 . Please re-open the issue if it's not resolved.
Thanks!
